xls/contrib/mlir/tools/xls_translate/xls_translate_to_mlir.cc

```diff
@@ -0,0 +1,1708 @@
+  }
+
+  auto opr_rhs =
+      state.getMlirValue(node.operands()[::xls::ArithOp::kRhsOperand]->id());
+  if (!opr_rhs.ok()) {
+    return opr_rhs.status();
+  }
+
+  auto result_type = translateType(node.GetType(), builder, ctx);
+  auto loc = translateLoc(node.loc(), builder, state);
+  if (!loc.ok()) {
+    return loc.status();
+  }
+
+  switch (node.op()) {
+    case ::xls::Op::kUMul:
+      return builder.create<xls::UmulOp>(*loc, result_type, *opr_lhs, *opr_rhs);
+    case ::xls::Op::kSMul:
+      return builder.create<xls::SmulOp>(*loc, result_type, *opr_lhs, *opr_rhs);
+    default:
+      return absl::InternalError(absl::StrCat(
+          "Expected ArithOp operation, not ", ::xls::OpToString(node.op())));
+  }
+}
+
+absl::StatusOr<Operation*> translateOp(::xls::PartialProductOp& node,
+                                       OpBuilder& builder, MLIRContext* ctx,
+                                       TranslationState& state) {
+  auto opr_lhs = state.getMlirValue(
+      node.operands()[::xls::PartialProductOp::kLhsOperand]->id());
+  if (!opr_lhs.ok()) {
+    return opr_lhs.status();
+  }
+
+  auto opr_rhs = state.getMlirValue(
+      node.operands()[::xls::PartialProductOp::kRhsOperand]->id());
+  if (!opr_rhs.ok()) {
+    return opr_rhs.status();
+  }
+
+  auto result_type = translateType(
+      node.GetType()->AsTupleOrDie()->element_type(0), builder, ctx);
+  auto loc = translateLoc(node.loc(), builder, state);
+  if (!loc.ok()) {
+    return loc.status();
+  }
+
+  Operation* op = nullptr;
+  switch (node.op()) {
+    case ::xls::Op::kSMulp:
+      op = builder.create<xls::SmulpOp>(*loc, result_type, *opr_lhs, *opr_rhs);
+      break;
+    case ::xls::Op::kUMulp:
+      op = builder.create<xls::UmulpOp>(*loc, result_type, *opr_lhs, *opr_rhs);
+      break;
+    default:
+      return absl::InternalError(
+          absl::StrCat("Expected PartialProductOp operation, not ",
+                       ::xls::OpToString(node.op())));
+  }
+
+  std::vector<Value> result_tuple_elems{};
+  result_tuple_elems.push_back(op->getResult(0));
+  result_tuple_elems.push_back(op->getResult(1));
+  ValueRange result_tuple(result_tuple_elems);
+  return builder.create<xls::TupleOp>(*loc, result_tuple);
+}
+
+absl::StatusOr<Operation*> translateOp(::xls::CompareOp& node,
+                                       OpBuilder& builder, MLIRContext* ctx,
+                                       TranslationState& state) {
+  auto opr_lhs =
+      state.getMlirValue(node.operands()[::xls::CompareOp::kLhsOperand]->id());
+  if (!opr_lhs.ok()) {
+    return opr_lhs.status();
+  }
+
+  auto opr_rhs =
+      state.getMlirValue(node.operands()[::xls::CompareOp::kRhsOperand]->id());
+  if (!opr_rhs.ok()) {
+    return opr_rhs.status();
+  }
+
+  auto result_type = translateType(node.GetType(), builder, ctx);
+  auto loc = translateLoc(node.loc(), builder, state);
+  if (!loc.ok()) {
+    return loc.status();
+  }
+
+  switch (node.op()) {
```

Comment:
There are sexier ways of doing this using TypeSwitch, but let's keep this as-is for now.

Reply:
Oh damn of course you're right. There's always std::visit, but frankly that might be grimmer (and is *awful* for compile time). Let's leave it.